File: Python/Caching/firstInFirstOut.py
from queue import Queue
def pageFaults(pages, n, capacity):
	s = set()
	indexes = Queue()
	page_faults = 0
	for i in range(n):
		if (len(s) < capacity):
			if (pages[i] not in s):
				s.add(pages[i])
				page_faults += 1
				indexes.put(pages[i])
		else:
			if (pages[i] not in s):
				val = indexes.queue[0]
				indexes.get()
				s.remove(val)
				s.add(pages[i])
				indexes.put(pages[i])
				page_faults += 1

	return page_faults

print("Enter the length of array: ")
n=int(input())
pages=[]
print("Enter the elements of array")
for i in range(n):
    pages.append(int(input()))
print("Enter the capacity")
capacity=int(input())
print("Answer:")
print(pageFaults(pages, n, capacity))

File: C++/Caching/firstInFirstOut.cpp
#include <iostream>
#include <queue>
#include <unordered_set>
using namespace std;

int pageFaults(int pages[], int n, int capacity)
{
    unordered_set<int> s;
    queue<int> indexes;
    int page_faults = 0;
    for (int i = 0; i < n; i++)
    {
        if (s.size() < capacity)
        {
            if (s.find(pages[i]) == s.end())
            {
                s.insert(pages[i]);
                page_faults++;
                indexes.push(pages[i]);
            }
        }
        else
        {
            if (s.find(pages[i]) == s.end())
            {
                int val = indexes.front();
                indexes.pop();
                s.erase(val);
                s.insert(pages[i]);
                indexes.push(pages[i]);
                page_faults++;
            }
        }
    }

    return page_faults;
}

int main()
{
    int n;
    cout << "Enter the length of the array: ";
    cin >> n;
    int pages[n];
    cout << "Enter the elements of the array: " << endl;
    for (int i = 0; i < n; i++)
    {
        cin >> pages[i];
    }
    int capacity;
    cout << "Enter the capacity: ";
    cin >> capacity;

    cout << "Answer: " << pageFaults(pages, n, capacity) << endl;

    return 0;
}
